lib/con_cache/lock_supervisor.ex

```diff
@@ -10,6 +10,8 @@ defmodule ConCache.LockSupervisor do
   end
 
   def start_link(n_partitions) do
+    true = Code.ensure_loaded?(ConCache.Lock.Resource)
```

Comment:
My guess is you're only experiencing slowdown in development (mode=interactive), in production releases (mode=embedded), all the modules would be preloaded on boot so this line would not be necessary. Btw, I vaguely remember that function might be going through the code server which we should rather avoid at runtime, though booting supervisors and similar is about the only place where this is sensible enough.

Interestingly enough, I believe this:

```elixir
  true = Code.ensure_loaded?(ConCache.Lock.Resource)

  def start_link(n_partitions) do
```

would solve your issue without impacting the runtime characteristics. That being said, ideally we fix the root cause, maybe it's code server and/or VM issue?

Reply:
Thanks Wojtek, pretty sure I’ve tried that before to no avail (as well as explicit require) but will double check

Reply:
Are you able to reproduce the slowdown with just con_cache? Ideally something like:

```
$ time elixir -e 'Mix.install([:con_cache]) ; {:ok, _} = ConCache.start_link(ttl_check_interval: false)'
```

Reply:
Sort of:

```elixir
iex(1)> ConCache.start_link(name: :sessions, ttl_check_interval: :timer.seconds(1), global_ttl: :timer.minutes(30))          
{:ok, #PID<0.215.0>}
iex(2)> :timer.tc(fn -> ConCache.isolated(:sessions, {:foo, :bar}, 500, fn -> :ok end) end)                                  
{5044, :ok}
iex(3)> :timer.tc(fn -> ConCache.isolated(:sessions, {:foo, :bar}, 500, fn -> :ok end) end)
{50, :ok}
```

and on this branch:

```elixir
iex(1)> ConCache.start_link(name: :sessions, ttl_check_interval: :timer.seconds(1), global_ttl: :timer.minutes(30))
{:ok, #PID<0.242.0>}
iex(2)> :timer.tc(fn -> ConCache.isolated(:sessions, {:foo, :bar}, 500, fn -> :ok end) end)                        
{2138, :ok}
iex(3)> :timer.tc(fn -> ConCache.isolated(:sessions, {:foo, :bar}, 500, fn -> :ok end) end)
{49, :ok}
```

Now with more processes:

```elixir
iex(1)> ConCache.start_link(name: :sessions, ttl_check_interval: :timer.seconds(1), global_ttl: :timer.minutes(30))                                 
{:ok, #PID<0.215.0>}
iex(2)> 1..System.schedulers_online() |> Task.async_stream(fn _ -> :timer.tc(fn -> ConCache.isolated(:sessions, {:foo, :bar}, 500, fn -> :ok end) end) end) |> Enum.take(24)
[
  ok: {6024, :ok},
  ok: {6083, :ok},
  ok: {6042, :ok},
  ok: {4637, :ok},
  ok: {4575, :ok},
  ok: {4529, :ok},
  ok: {4516, :ok},
  ok: {4481, :ok},
  ok: {4451, :ok},
  ok: {4434, :ok},
  ok: {4426, :ok},
  ok: {4380, :ok},
  ok: {4366, :ok},
  ok: {4355, :ok},
  ok: {4367, :ok},
  ok: {4363, :ok},
  ok: {4352, :ok},
  ok: {4347, :ok},
  ok: {4337, :ok},
  ok: {4334, :ok},
  ok: {4320, :ok},
  ok: {4311, :ok},
  ok: {4304, :ok},
  ok: {4273, :ok}
]
```

and on this branch:

```elixir
iex(1)> ConCache.start_link(name: :sessions, ttl_check_interval: :timer.seconds(1), global_ttl: :timer.minutes(30))                                                         
{:ok, #PID<0.215.0>}
iex(2)> 1..System.schedulers_online() |> Task.async_stream(fn _ -> :timer.tc(fn -> ConCache.isolated(:sessions, {:foo, :bar}, 500, fn -> :ok end) end) end) |> Enum.take(24)
[
  ok: {2935, :ok},
  ok: {2984, :ok},
  ok: {2969, :ok},
  ok: {2532, :ok},
  ok: {2577, :ok},
  ok: {2566, :ok},
  ok: {2542, :ok},
  ok: {2514, :ok},
  ok: {2525, :ok},
  ok: {2544, :ok},
  ok: {2533, :ok},
  ok: {2509, :ok},
  ok: {2485, :ok},
  ok: {2421, :ok},
  ok: {2386, :ok},
  ok: {2372, :ok},
  ok: {2351, :ok},
  ok: {2316, :ok},
  ok: {2292, :ok},
  ok: {2264, :ok},
  ok: {2250, :ok},
  ok: {2224, :ok},
  ok: {2192, :ok},
  ok: {2189, :ok}
]
```

Subsequent call:

```elixir
iex(3)> 1..System.schedulers_online() |> Task.async_stream(fn _ -> :timer.tc(fn -> ConCache.isolated(:sessions, {:foo, :bar}, 500, fn -> :ok end) end) end) |> Enum.take(24)
[
  ok: {71, :ok},
  ok: {42, :ok},
  ok: {16, :ok},
  ok: {47, :ok},
  ok: {75, :ok},
  ok: {57, :ok},
  ok: {66, :ok},
  ok: {49, :ok},
  ok: {47, :ok},
  ok: {38, :ok},
  ok: {45, :ok},
  ok: {97, :ok},
  ok: {94, :ok},
  ok: {131, :ok},
  ok: {162, :ok},
  ok: {164, :ok},
  ok: {140, :ok},
  ok: {134, :ok},
  ok: {142, :ok},
  ok: {85, :ok},
  ok: {108, :ok},
  ok: {98, :ok},
  ok: {58, :ok},
  ok: {55, :ok}
]
```

Reply:
I'm guessing there's more to do if we care about avoiding the code server in interactive mode. I've tried running https://github.com/DockYard/flame_on and it was indeed the culprit (hence the patch) - apart from the awkward `MapSet` instantiation.

Reply:
I wonder if this happens because of concurrent attempts to load the module. 

Looking at the code of `ConCache.Resource`, the module is short, and there's no compile-time generation of functions and clauses, so this all seems very strange.

This needs further investigation, and I'm currently not inclined to add this workaround to ConCache. We should instead try to check if the similar problem exists in a simplified small project, where we try to do concurrent implicit lazy load of some small module.